spec/std/uri_spec.cr

```diff
@@ -305,112 +346,96 @@ describe "URI" do
+    it_encodes("hello%2+world", "hello%252+world")
+    it_encodes("なな", "%E3%81%AA%E3%81%AA")
+    it_encodes("'Stop!' said Fred", "'Stop!'%20said%20Fred")
+    it_encodes("\n", "%0A")
+    it_encodes("https://en.wikipedia.org/wiki/Crystal (programming language)", "https://en.wikipedia.org/wiki/Crystal%20(programming%20language)")
+    it_encodes("\xFF", "%FF") # encodes invalid UTF-8 character
+    it_encodes("hello world", "hello+world", space_to_plus: true)
+    it_encodes("'Stop!' said Fred", "'Stop!'+said+Fred", space_to_plus: true)
+
+    it "does not encode character when block returns true" do
+      String.build do |io|
+        URI.decode("hello&world", io) { |byte| URI.reserved? byte }
+      end.should eq("hello&world")
     end
+  end
 
-    it "escape space to plus when space_to_plus flag is true" do
-      URI.escape("hello world", space_to_plus: true).should eq("hello+world")
-      URI.escape("'Stop!' said Fred", space_to_plus: true).should eq("%27Stop%21%27+said+Fred")
-    end
+  describe ".encode_www_form" do
+    it_encodes_www_form("", "")
+    it_encodes_www_form("abc", "abc")
+    it_encodes_www_form("1%41", "1%2541")
+    it_encodes_www_form("a b+", "a+b%2B")
+    it_encodes_www_form("a b+", "a%20b%2B", space_to_plus: false)
+    it_encodes_www_form("10%", "10%25")
+    it_encodes_www_form(" ?&=#+%!<>#\"{}|\\^[]`☺\t:/@$'()*,;", "+%3F%26%3D%23%2B%25%21%3C%3E%23%22%7B%7D%7C%5C%5E%5B%5D%60%E2%98%BA%09%3A%2F%40%24%27%28%29%2A%2C%3B")
+    it_encodes_www_form("* foo=bar baz&hello/", "%2A+foo%3Dbar+baz%26hello%2F")
+  end
 
-    it "does not escape character when block returns true" do
-      URI.unescape("hello&world") { |byte| URI.reserved? byte }
-        .should eq("hello&world")
-    end
+  describe ".decode_www_form" do
+    it_decodes_www_form("", "")
+    it_decodes_www_form("abc", "abc")
+    it_decodes_www_form("1%41", "1A")
+    it_decodes_www_form("1%41%42%43", "1ABC")
+    it_decodes_www_form("%4a", "J")
+    it_encodes_www_form("hello+", "hello%2B")
+    it_encodes_www_form("hello+world", "hello%2Bworld")
+    it_encodes_www_form("hello%2+world", "hello%252%2Bworld")
+    it_encodes_www_form("'Stop!' said Fred", "%27Stop%21%27+said+Fred")
+    it_decodes_www_form("a+b", "a b")
+    it_decodes_www_form("a%20b", "a b")
+    it_decodes_www_form("%20%3F%26%3D%23%2B%25%21%3C%3E%23%22%7B%7D%7C%5C%5E%5B%5D%60%E2%98%BA%09%3A%2F%40%24%27%28%29%2A%2C%3B", " ?&=#+%!<>#\"{}|\\^[]`☺\t:/@$'()*,;")
+    it_decodes_www_form("+%2B %20", " +  ")
+
+    it_decodes_www_form("%", "%")
+    it_decodes_www_form("%1", "%1")
+    it_decodes_www_form("123%45%6", "123E%6")
+    it_decodes_www_form("%zzzzz", "%zzzzz")
   end
 
-  describe "reserved?" do
-    reserved_chars = Set.new([':', '/', '?', '#', '[', ']', '@', '!', '$', '&', '\'', '(', ')', '*', '+', ',', ';', '='])
+  it ".reserved?" do
+    reserved_chars = {':', '/', '?', '#', '[', ']', '@', '!', '$', '&', '\'', '(', ')', '*', '+', ',', ';', '='}
```

Comment:
I don't think there's a need to use a tuple here.

Reply:
That is, each different tuple type means a new type the compiler has to track, with new method instantiations and so on. So if the idea of using a tuple is for performance, this actually had the opposite result. And in specs the performance of using a tuple vs a set doesn't matter (and I think a Set is more efficient for lookup).